core/src/main/java/software/amazon/awssdk/retry/RetryHandler.java

```diff
@@ -0,0 +1,143 @@
+    private boolean retryCapacityConsumed;
+    private RetryPolicyContext retryPolicyContext;
+    private Optional<SdkBaseException> lastRetriedException = Optional.empty();
+
+    public RetryHandler(RetryPolicy retryPolicy,
+                        CapacityManager retryCapacity) {
+        this.retryPolicy = retryPolicy;
+        this.retryCapacity = retryCapacity;
+    }
+
+    public boolean shouldRetry(HttpResponse httpResponse,
+                               SdkHttpFullRequest request,
+                               RequestExecutionContext context,
+                               SdkBaseException exception,
+                               int requestCount) {
+
+        final int retriesAttempted = requestCount - 1;
+
+        // Do not use retry capacity for throttling exceptions
+        if (!RetryUtils.isThrottlingException(exception)) {
+            // See if we have enough available retry capacity to be able to execute this retry attempt.
+            if (!retryCapacity.acquire(RetryPolicy.THROTTLED_RETRY_COST)) {
+                return false;
+            }
+            this.retryCapacityConsumed = true;
+        }
+
+        this.retryPolicyContext = RetryPolicyContext.builder()
+                                                    .request(request)
+                                                    .originalRequest(context.requestConfig().getOriginalRequest())
+                                                    .exception(exception)
+                                                    .retriesAttempted(retriesAttempted)
+                                                    .httpStatusCode(httpResponse == null ? null : httpResponse.getStatusCode())
+                                                    .build();
+        // Finally, pass all the context information to the RetryCondition and let it decide whether it should be retried.
+        if (!retryPolicy.shouldRetry(retryPolicyContext)) {
+            // If the retry policy fails we immediately return consumed capacity to the pool.
+            if (retryCapacityConsumed) {
+                retryCapacity.release(RetryPolicy.THROTTLED_RETRY_COST);
+            }
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * If this was a successful retry attempt we'll release the full retry capacity that the attempt originally consumed.  If
+     * this was a successful initial request we release a lesser amount.
+     */
+    public void releaseRetryCapacity() {
+        if (isRetry() && retryCapacityConsumed) {
+            retryCapacity.release(RetryPolicy.THROTTLED_RETRY_COST);
+        } else {
+            retryCapacity.release();
+        }
+    }
+
+    /**
+     * Computes the delay before the next retry should be attempted based on the retry policy context.
+     * @return long value of how long to wait
+     */
+    public long computeDelayBeforeNextRetry() {
+        long delay = retryPolicy.computeDelayBeforeNextRetry(retryPolicyContext);
```

Comment:
we don't need local variable 'delay' here. why not directly assign to lastBackoffDelay and return it?